sigstore/dsse.py

```diff
@@ -0,0 +1,211 @@
+
+    This type deals with opaque bytes to ensure that the encoding does not
+    change, but Statements are internally checked for conformance against
+    the JSON object layout defined in the in-toto attesation spec.
+
+    See: <https://github.com/in-toto/attestation/blob/main/spec/v1/statement.md>
+    """
+
+    def __init__(self, contents: bytes) -> None:
+        """
+        Construct a new Statement.
+
+        This takes an opaque `bytes` containing the statement; use
+        `StatementBuilder` to manually construct an in-toto statement
+        from constituent pieces.
+        """
+        self._contents = contents
+        try:
+            self._statement = _Statement.model_validate_json(contents)
+        except ValidationError:
+            raise ValueError("malformed in-toto statement")
+
+    def _pae(self) -> bytes:
+        """
+        Construct the PAE encoding for this statement.
+        """
+
+        # See:
+        # https://github.com/secure-systems-lab/dsse/blob/v1.0.0/envelope.md
+        # https://github.com/in-toto/attestation/blob/v1.0/spec/v1.0/envelope.md
+        pae = f"DSSEv1 {len(Envelope._TYPE)} {Envelope._TYPE} ".encode()
+        pae += b" ".join([str(len(self._contents)).encode(), self._contents])
+        return pae
+
+
+class _StatementBuilder:
+    """
+    A builder-style API for constructing in-toto Statements.
+    """
+
+    def __init__(
+        self,
+        subjects: Optional[List[_Subject]] = None,
+        predicate_type: Optional[str] = None,
+        predicate: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Create a new `_StatementBuilder`.
+        """
+        self._subjects = subjects or []
+        self._predicate_type = predicate_type
+        self._predicate = predicate
+
+    def subjects(self, subjects: list[_Subject]) -> _StatementBuilder:
+        """
+        Configure the subjects for this builder.
+        """
+        self._subjects = subjects
+        return self
+
+    def predicate_type(self, predicate_type: str) -> _StatementBuilder:
+        """
+        Configure the predicate type for this builder.
+        """
+        self._predicate_type = predicate_type
+        return self
+
+    def predicate(self, predicate: dict[str, Any]) -> _StatementBuilder:
+        """
+        Configure the predicate for this builder.
+        """
+        self._predicate = predicate
+        return self
+
+    def build(self) -> Statement:
+        """
+        Build a `Statement` from the builder's state.
+        """
+        try:
+            stmt = _Statement(
+                type_="https://in-toto.io/Statement/v1",
+                subjects=self._subjects,
+                predicate_type=self._predicate_type,
+                predicate=self._predicate,
+            )
+        except ValidationError as e:
+            raise ValueError(f"invalid statement: {e}")
+
+        return Statement(stmt.model_dump_json(by_alias=True).encode())
+
+
+class Envelope:
```

Comment:
Added this as a thin wrapper for `sigstore_protobuf_specs.io.intoto.Envelope`, so that we don't leak a protobuf model into the public API here.